### PR TITLE
Implement DepositProxy logic

### PR DESCRIPTION
**Pull request by haoze-grvt**
Stacked PRs:
 * __->__#222

--- --- ---

### CODE REVIEW

File: hardhat.config.ts [hunk 1]
```diff
@@ -51,6 +51,7 @@ const config: HardhatUserConfig = {
   zksolc: {
     version: "1.5.1",
     settings: {
+      isSystem: true,
```
Comment (minhbsq) [inline hardhat.config.ts:54]: what is this flag?

File: contracts/exchange/GRVTExchange.sol [hunk 2]
```diff
@@ -22,7 +24,13 @@ contract GRVTExchange is
     _disableInitializers();
   }
 
-  function initialize(address admin, address chainSubmitter, address initializeConfigSigner) public initializer {
+  function initialize(
+    address admin,
+    address chainSubmitter,
+    address initializeConfigSigner,
+    address beaconOwner,
+    bytes32 depositProxyProxyBytecodeHash
```
Comment (minhbsq) [inline contracts/exchange/GRVTExchange.sol:32]: is `ProxyProxy` name intentional?

File: contracts/exchange/GRVTExchange.sol [hunk 3]
```diff
@@ -22,7 +24,13 @@ contract GRVTExchange is
     _disableInitializers();
   }
 
-  function initialize(address admin, address chainSubmitter, address initializeConfigSigner) public initializer {
+  function initialize(
```
Comment (minhbsq) [inline contracts/exchange/GRVTExchange.sol:27]: would be good to add a comment explaining each of the variable passed in

File: contracts/exchange/api/BaseContract.sol [hunk 4]
```diff
@@ -337,4 +349,45 @@ contract BaseContract is AccessControlUpgradeable, ReentrancyGuardUpgradeable {
 
     return pos;
   }
+
+  function _deployDepositProxy(address account) internal returns (address) {
```
Comment (minhbsq) [inline contracts/exchange/api/BaseContract.sol:353]: Assume this function will be read by auditors, let's add comments explaining this function here

File: contracts/exchange/api/BaseContract.sol [hunk 5]
```diff
@@ -337,4 +349,45 @@ contract BaseContract is AccessControlUpgradeable, ReentrancyGuardUpgradeable {
...
+
+    return depositProxy;
+  }
+
+  function _getCreate2Salt(address accountID) internal pure returns (bytes32 salt) {
+    salt = bytes32(uint256(uint160(accountID)));
+  }
+
+  /// @dev Deploy the beacon proxy for the L2 token, while using ContractDeployer system contract.
+  /// @dev This function uses raw call to ContractDeployer to make sure that exactly `depositProxyProxyBytecodeHash` is used
```
Comment (minhbsq) [inline contracts/exchange/api/BaseContract.sol:367]: i don't understand this part, can you explain?

File: hardhat.config.ts (hunk 1 above)
Comment (haoze-grvt) [reply hardhat.config.ts:54]: This is to enables Yul instructions available only for ZKsync system contracts. We need this flag to be able to deploy contracts from contract. I checked with ML before and there is no concern enabling this flag.

File: contracts/exchange/GRVTExchange.sol (hunk 2 above)
Comment (haoze-grvt) [reply contracts/exchange/GRVTExchange.sol:32]: Yes, it's the `BeaconProxy` for the `DepositProxy`.

File: contracts/exchange/api/BaseContract.sol (hunk 5 above)
Comment (haoze-grvt) [reply contracts/exchange/api/BaseContract.sol:367]: Updated comment

File: contracts/exchange/api/BaseContract.sol (hunk 4 above)
Comment (haoze-grvt) [reply contracts/exchange/api/BaseContract.sol:353]: Added comment

File: contracts/exchange/GRVTExchange.sol (hunk 3 above)
Comment (haoze-grvt) [reply contracts/exchange/GRVTExchange.sol:27]: Added comment

File: hardhat.config.ts (hunk 1 above)
Comment (minhbsq) [reply hardhat.config.ts:54]: I see now

**Review by minhbsq — APPROVED** [review]
ship it revert

**Pull request merged**